### PR TITLE
Feature/co 1033 integration tests

### PR DESCRIPTION
**Pull request by jaredkosanovic**

### CODE REVIEW

**Review by tsoliangwu0130 — CHANGES_REQUESTED** [review]
1. The JIRA ticket number should be CO-1033
2. Don't forget to add `README.md` for the integration test
3. Could you add `configuration.json` to `.dockerignore` as well?

File: src/test/integration/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,14 @@
+FROM python:3
+
+COPY requirements.txt ./
+RUN pip install --no-cache-dir -r requirements.txt
```
Comment (tsoliangwu0130) [inline src/test/integration/Dockerfile:4]: Just curious, did you run into any issue which makes you decide to disable the cache here?

File: src/test/integration/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+WORKDIR /usr/src/app
+
+RUN ["chmod", "o+w", "."]
```
Comment (tsoliangwu0130) [inline src/test/integration/Dockerfile]: why do you need to add write permission here?

File: src/test/integration/integration_tests.py [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        return random_datetime.date()
+
+    def test_all_locations(self):
+        request = self.__locations_request(None)
+        request_elapsed_seconds = request.elapsed.total_seconds()
+        logging.debug("Request took {} second(s)"
+                      .format(request_elapsed_seconds))
+
+        self.assertEqual(request.status_code, 200)
+        self.assertLess(request_elapsed_seconds, 1)
```
Comment (tsoliangwu0130) [inline src/test/integration/integration_tests.py]: I ran the integration test multiple times. It seems like this assertion is not always true:

```
test_all_locations (__main__.integration_tests) ... FAIL
test_date_parameter (__main__.integration_tests) ... ok
test_departments (__main__.integration_tests) ... ok
test_departments_bad_requests (__main__.integration_tests) ... ok
test_position_type_case (__main__.integration_tests) ... ok
test_positions (__main__.integration_tests) ... ok
test_positions_bad_requests (__main__.integration_tests) ... ok
test_state_parameter (__main__.integration_tests) ... ok

======================================================================
FAIL: test_all_locations (__main__.integration_tests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "./integration_tests.py", line 189, in test_all_locations
    self.assertLess(request_elapsed_seconds, 1)
AssertionError: 2.630726 not less than 1

----------------------------------------------------------------------
Ran 8 tests in 10.811s

FAILED (failures=1)
```

File: src/test/integration/integration_tests.py [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        client_id = config_file_json["client_id"]
+        client_secret = config_file_json["client_secret"]
+
+        access_token = cls.get_access_token(
+                config_file_json["token_api"],
+                client_id,
+                client_secret)
+
+        # Set headers and query parameters
+        cls.auth_header = {'Authorization': 'Bearer %s' % access_token}
```
Comment (tsoliangwu0130) [inline src/test/integration/integration_tests.py]: In your code, you used  both `'%s %s' % ('one', 'two')` and `'{} {}'.format('one', 'two')` two different style for the string formatting. I will recommend you stick with one of the style for the code consistency.

Reference: https://pyformat.info/

File: src/test/integration/integration_tests.py [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        request = self.__positions_request(self.business_centers[0],
+                                           self.position_type.upper())
+        self.assertEqual(request.status_code, 200)
+
+    def test_positions_bad_requests(self):
+        self.assertEqual(self.__positions_request(
+                         None, self.position_type).status_code, 400)
+        self.assertEqual(self.__positions_request(
+                         "Bad bc", self.position_type).status_code, 400)
+        self.assertEqual(self.__positions_request(
```
Comment (tsoliangwu0130) [inline src/test/integration/integration_tests.py:150]: Could you also add one test case which `type` parameter is `None` like you did at [line 148](https://github.com/osu-mist/hr-api/blob/d0c624a3bf4176d0fdcbb266eca49024c354f4d5/src/test/integration/integration_tests.py#L148)?

File: src/test/integration/integration_tests.py [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+        for bc in self.business_centers:
+            logging.debug("Testing business center: {}".format(bc))
+            request = self.__departments_request(bc)
+
+            request_elapsed_seconds = request.elapsed.total_seconds()
+            logging.debug("Request took {} second(s)"
+                          .format(request_elapsed_seconds))
+
+            self.assertEqual(request.status_code, 200)
+            self.assertLess(request_elapsed_seconds, 1)
```
Comment (tsoliangwu0130) [inline src/test/integration/integration_tests.py]: Since you tested the elapsed seconds for every request, could you put this test case to another function to make this logic reusable?

**Review by tsoliangwu0130 — CHANGES_REQUESTED** [review]

File: src/test/integration/Dockerfile (hunk 1 above)
Comment (jaredkosanovic) [reply src/test/integration/Dockerfile:4]: No good reason, I think I reused this file from another integration test. My naive answer is that we don't need to cache anything from pip in the docker image.

File: src/test/integration/Dockerfile (hunk 2 above)
Comment (jaredkosanovic) [reply src/test/integration/Dockerfile]: I guess I don't. I reused this from another integration test.

File: src/test/integration/integration_tests.py (hunk 3 above)
Comment (jaredkosanovic) [reply src/test/integration/integration_tests.py]: I was probably testing enough that the backend datasource was caching things. You're right that this test will fail if it's run for the first time in a while. I will increase the tolerance to 3 seconds.

File: src/test/integration/integration_tests.py (hunk 4 above)
Comment (jaredkosanovic) [reply src/test/integration/integration_tests.py]: Good catch.

File: src/test/integration/integration_tests.py (hunk 5 above)
Comment (jaredkosanovic) [reply src/test/integration/integration_tests.py:150]: Done.
